archive/imo/imo2021_q1.lean

```diff
@@ -0,0 +1,273 @@
+  --argument from lemma above.
+  by_cases p : n ≥ 107,
+  { have h := exists_numbers_in_interval n p,
+    cases h with l hl,
+    use (2 * l * l - 4 * l),
+    use (2 * l * l + 1),
+    use (2 * l * l + 4 * l),
+    by_cases p : l > 1,
+    refine ⟨_,⟨_,⟨_,⟨_,⟨_, ⟨_, _⟩⟩⟩⟩⟩⟩,
+    { exact nat.le_sub_right_of_add_le hl.1 },
+    { exact (2 * l * l).sub_lt_succ (4 * l) },
+    { linarith },
+    { exact hl.2 },
+    { use (2 * l - 1),
+      have h₁ : 1 ≤ 2 * l := by linarith,
+      have h₂ : 4 * l ≤ 2 * l * l := by linarith,
+      zify [h₁, h₂],
+      ring_exp_eq },
+    { use (2 * l),
+      have h₁ : 4 * l ≤ 2 * l * l := by linarith,
+      zify [h₁],
+      ring_exp_eq },
+    { use(2 * l + 1),
+      ring_exp_eq,
+    },
+      --Otherwise, if 100 ≤ n < 107, then it suffices to consider explicit
+      --construction of a triplet {a, b, c}, which is constructed by setting l=9
+      --in the argument at the start of the file.
+    { exfalso,
+      simp only [not_lt] at p,
+      interval_cases l,
+      all_goals {linarith }}},
+  { refine ⟨126, 163, 198, _, _, _, _, ⟨⟨17, _⟩,⟨⟨18, _⟩,⟨19, _⟩⟩⟩⟩,
+    all_goals { linarith }},
+end
+
+lemma mem_compl_iff_nmem_subset {A : finset ℕ} {a b c n : ℕ}
+(t : A ⊆ finset.Ico n (2 * n + 2)) (h₁ : n ≤ a)
+(h₂ : a < b) (h₃ : b < c) (h₄ : c ≤ 2 * n) :
+(a ∉ A ↔ a ∈ finset.Ico n (2 * n + 2) \ A) ∧
+(b ∉ A ↔ b ∈ finset.Ico n (2 * n + 2) \ A) ∧
+(c ∉ A ↔ c ∈ finset.Ico n (2 * n + 2) \ A) :=
+begin
+  refine ⟨⟨_,_⟩, ⟨⟨_,_⟩,⟨_,_⟩⟩⟩,
+  { intro ha,
+    simp only [finset.Ico.mem, finset.mem_sdiff],
+    refine ⟨⟨_,_⟩,ha⟩,
+    all_goals { linarith }},
+  { intro ha,
+    simp only [finset.mem_sdiff, finset.mem_range] at ha,
+    exact ha.2 },
+  { intro hb,
+    simp only [finset.Ico.mem, finset.mem_sdiff],
+    refine ⟨⟨_,_⟩,hb⟩,
+    all_goals { linarith }},
+  { intro hb,
+    simp only [finset.mem_sdiff, finset.mem_range] at hb,
+    exact hb.2 },
+  { intro hc,
+    simp only [finset.Ico.mem, finset.mem_sdiff],
+    refine ⟨⟨_,_⟩,hc⟩,
+    all_goals { linarith }},
+  { intro hc,
+    simp only [finset.mem_sdiff, finset.mem_range] at hc,
+    exact hc.2 },
+end
+
+
+theorem IMO_2021_Q1 : ∀ (n : ℕ), n ≥ 100 → ∀ (A ⊆ finset.Ico n (2 * n + 2)),
+  (∃ (a b : A), a ≠ b ∧ ∃ (k : ℕ), (a:ℕ) + b = k * k) ∨
+  (∃ (a b : finset.Ico n (2 * n + 2) \ A), a ≠ b ∧ ∃(k : ℕ), (a:ℕ) + b = k * k) :=
+begin
+  intros n hn A hA,
+  -- There exists a pairwise unequal triplet a, b, c ∈ [n, 2n]
+  -- such that all pairwise sums are perfect squares.
+  have p := exists_triplet_summing_to_squares n hn,
+  rcases p with ⟨a, b, c, ⟨han, hba, hcb, hnc, p⟩⟩,
+  -- Consider by cases based on whether each of {a, b, c} belongs to the 'first pile' A or not.
+  -- In each case, we can find two members of the triplet in the same pile, as required.
+  by_cases h₁: a ∈ A,
+  { by_cases h₂: b ∈ A,
+    { left,
+      use a,
+      exact h₁,
+      use b,
+      exact h₂,
```

Comment:
This looks like there is more than one goal open. Please use `{}` to organize the proof so that there is only 1 goal open at a time.